datadogreceiver/translator.go

```diff
@@ -0,0 +1,166 @@
+	rspan := resSpans.AppendEmpty()
+	resource := rspan.Resource()
+	addResourceData(req, &resource)
+
+	ils := rspan.ScopeSpans().AppendEmpty()
+
+	ils.Scope().SetName("Datadog-" + req.Header.Get("Datadog-Meta-Lang"))
+	ils.Scope().SetVersion(req.Header.Get("Datadog-Meta-Tracer-Version"))
+
+	for _, trace := range traces {
+
+		spans := ptrace.NewSpanSlice()
+		spans.EnsureCapacity(len(trace))
+		for _, span := range trace {
+			newSpan := spans.AppendEmpty()
+			newSpan.SetTraceID(uInt64ToTraceID(0, span.TraceID))
+			newSpan.SetSpanID(uInt64ToSpanID(span.SpanID))
+			newSpan.SetStartTimestamp(pcommon.Timestamp(span.Start))
+			newSpan.SetEndTimestamp(pcommon.Timestamp(span.Start + span.Duration))
+			newSpan.SetParentSpanID(uInt64ToSpanID(span.ParentID))
+			newSpan.SetName(span.Resource)
+
+			if span.Error > 0 {
+				newSpan.Status().SetCode(ptrace.StatusCodeError)
+			} else {
+				newSpan.Status().SetCode(ptrace.StatusCodeOk)
+			}
+
+			attrs := newSpan.Attributes()
+			attrs.EnsureCapacity(len(span.GetMeta()) + 1)
+			attrs.InsertString(semconv.AttributeServiceName, span.Service)
+			for k, v := range span.GetMeta() {
+				k = translateDataDogKeyToOtel(k)
+				if len(k) > 0 {
+					attrs.InsertString(k, v)
+				}
+			}
+
+			switch span.Type {
+			case "web":
+				newSpan.SetKind(ptrace.SpanKindServer)
+			case "custom":
+				newSpan.SetKind(ptrace.SpanKindUnspecified)
+			default:
+				newSpan.SetKind(ptrace.SpanKindClient)
+			}
+		}
+		spans.MoveAndAppendTo(ils.Spans())
+	}
+
+	return dest
+}
+
+func translateDataDogKeyToOtel(k string) string {
+	// Tags prefixed with _dd. are for Datadog's use only, adding them to another backend will just increase cardinality needlessly
+	if strings.HasPrefix(k, "_dd.") {
+		return ""
+	}
+	switch strings.ToLower(k) {
+	case "env":
+		return semconv.AttributeDeploymentEnvironment
+	case "version":
+		return semconv.AttributeServiceVersion
+	case "container_id":
+		return semconv.AttributeContainerID
+	case "container_name":
+		return semconv.AttributeContainerName
+	case "image_name":
+		return semconv.AttributeContainerImageName
+	case "image_tag":
+		return semconv.AttributeContainerImageTag
+	case "process_id":
+		return semconv.AttributeProcessPID
+	case "error.stacktrace":
+		return semconv.AttributeExceptionStacktrace
+	case "error.msg":
+		return semconv.AttributeExceptionMessage
+	default:
+		return k
+	}
+
+}
+
+func decodeRequest(req *http.Request, dest *datadogpb.Traces) error {
+	switch mediaType := getMediaType(req); mediaType {
+	case "application/msgpack":
+		if strings.HasPrefix(req.URL.Path, "/v0.5") {
+			reader := datadogpb.NewMsgpReader(req.Body)
+			defer datadogpb.FreeMsgpReader(reader)
+			return dest.DecodeMsgDictionary(reader)
+		}
+		return msgp.Decode(req.Body, dest)
+	default:
+		return json.NewDecoder(req.Body).Decode(dest)
+	}
+}
+
+func getMediaType(req *http.Request) string {
+	mt, _, err := mime.ParseMediaType(req.Header.Get("Content-Type"))
```

Comment:
What does `mime.ParseMediaType` do? Is it just a case insensitive check?

Reply:
Supposedly content type may contains more than just the media type. The example on [https://pkg.go.dev/mime#ParseMediaType](https://pkg.go.dev/mime#ParseMediaType) shows charset is par tof the same header.

Reply:
Looking at what the dd agent headers. Im not sure what it would contain but my example just returned "application/msgpack" for Content-Type.  So not sure if its really needed. Perhaps the author ran into an issue otherwise?